lib/utils/compose.js

```diff
@@ -0,0 +1,1087 @@
+		});
+
+const authorizePush = function(
+	sdk,
+	tokenAuthEndpoint,
+	registry,
+	images,
+	previousRepos,
+) {
+	if (!Array.isArray(images)) {
+		images = [images];
+	}
+
+	images.push(...previousRepos);
+	return sdk.request
+		.send({
+			baseUrl: tokenAuthEndpoint,
+			url: '/auth/v1/token',
+			qs: {
+				service: registry,
+				scope: images.map(repo => `repository:${repo}:pull,push`),
+			},
+		})
+		.get('body')
+		.get('token')
+		.catchReturn({});
+};
+
+const pushAndUpdateServiceImages = function(docker, token, images, afterEach) {
+	const { DockerProgress } = require('docker-progress');
+	const { retry } = require('./helpers');
+	const tty = require('./tty')(process.stdout);
+
+	const opts = { authconfig: { registrytoken: token } };
+
+	const progress = new DockerProgress({ dockerToolbelt: docker });
+	const renderer = pushProgressRenderer(
+		tty,
+		getChalk().blue('[Push]') + '    ',
+	);
+	const reporters = progress.aggregateProgress(images.length, renderer);
+
+	return Promise.using(tty.cursorHidden(), () =>
+		Promise.map(images, ({ serviceImage, localImage, props, logs }, index) =>
+			Promise.join(
+				localImage.inspect().get('Size'),
+				retry(
+					() => {
+						progress.push(localImage.name, reporters[index], opts);
+					},
+					3, // `times` - retry 3 times
+					localImage.name, // `label` included in retry log messages
+					2000, // `delayMs` - wait 2 seconds before the 1st retry
+					1.4, // `backoffScaler` - wait multiplier for each retry
+				).finally(renderer.end),
+				function(size, digest) {
+					serviceImage.image_size = size;
+					serviceImage.content_hash = digest;
+					serviceImage.build_log = logs;
+					serviceImage.dockerfile = props.dockerfile;
+					serviceImage.project_type = props.projectType;
+					if (props.startTime) {
+						serviceImage.start_timestamp = props.startTime;
+					}
+					if (props.endTime) {
+						serviceImage.end_timestamp = props.endTime;
+					}
+					serviceImage.push_timestamp = new Date();
+					serviceImage.status = 'success';
+				},
+			)
+				.tapCatch(function(e) {
+					serviceImage.error_message = '' + e;
+					serviceImage.status = 'failed';
+				})
+				.finally(() => afterEach?.(serviceImage, props)),
+		),
+	);
+};
+
+export function deployProject(
+	docker,
+	logger,
+	composition,
+	images,
+	appId,
+	userId,
+	auth,
+	apiEndpoint,
+	skipLogUpload,
+) {
+	const _ = require('lodash');
+	const releaseMod = require('balena-release');
+	const tty = require('./tty')(process.stdout);
+
+	const prefix = getChalk().cyan('[Info]') + '    ';
+	const spinner = createSpinner();
+	let runloop = runSpinner(tty, spinner, `${prefix}Creating release...`);
+
+	return createRelease(apiEndpoint, auth, userId, appId, composition)
+		.finally(runloop.end)
+		.then(function({ client, release, serviceImages }) {
+			logger.logDebug('Tagging images...');
+			return tagServiceImages(docker, images, serviceImages)
+				.tap(function(taggedImages) {
```

Comment:
I renamed this from `images` to `taggedImages` to avoid a shadowed var linting issue